File: Interview-System Design Practice/Parking Lot/Python/parking_lot.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Parking lot module.
"""

from collections import deque
from enum import Enum, auto

from vehicle import Bus, Car, MotorCycle, Truck, Vehicle


class ParkingSpotSize(Enum):
    """
    Sizes of the parking spots.
    """
    S = auto()
    M = auto()
    L = auto()
    XL = auto()


class ParkingSpot:
    __slots__ = ['_id', '_size', '_my_vehicle']

    _next_id = 0

    def __init__(self, size: ParkingSpotSize):
        """
        Constructor with parameter.
        :param size: ParkingSpotSize
        """
        self._id = ParkingSpot._next_id
        ParkingSpot._next_id += 1
        self._size = size
        self._my_vehicle = None

    @property
    def id(self) -> int:
        """
        Accessor of id.
        :return: int
        """
        return self._id

    @property
    def size(self) -> ParkingSpotSize:
        """
        Accessor of size.
        :return: ParkingSpotSize
        """
        return self._size

    @property
    def vehicle(self) -> Vehicle:
        """
        Accessor of my_vehicle.
        :return: Vehicle
        """
        return self._my_vehicle

    @vehicle.setter
    def vehicle(self, vehicle: Vehicle) -> None:
        """
        Mutator of my_vehicle.
        :param vehicle: Vehicle
        :return: None
        """
        self._my_vehicle = vehicle


class ParkingLot:
    __slots__ = ['_addr', '_available', '_parked']

    def __init__(self, addr: str):
        """
        Constructor with parameter.
        :param addr: str
        """
        self._addr = addr
        self._available = {size: deque() for size in ParkingSpotSize}
        self._parked = {}

    def add_parking_spot(self, size: ParkingSpotSize) -> None:
        """
        Adds a new parking spot of the given size.
        :param size: ParkingSpotSize
        :return: None
        """
        size_available = self._parked[size]
        size_available.offer(ParkingSpot(size))
        self._parked[size] = size_available

    def place_vehicle(self, vehicle: Vehicle) -> int:
        """
        Places the given vehicle.
        :param vehicle: Vehicle
        :return: int
        """
        # Check whether the vehicle is already parked
        if vehicle.license_plate in self._parked:
            raise Exception('The vehicle is already parked.')

        if isinstance(vehicle, MotorCycle):
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.S)
            if parked_spot_id != -1:
                return parked_spot_id
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.M)
            if parked_spot_id != -1:
                return parked_spot_id
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.L)
            if parked_spot_id != -1:
                return parked_spot_id
            return self._place_vehicle_to_sized_spot(vehicle,
                                                     ParkingSpotSize.XL)
        elif isinstance(vehicle, Car):
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.M)
            if parked_spot_id != -1:
                return parked_spot_id
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.L)
            if parked_spot_id != -1:
                return parked_spot_id
            return self._place_vehicle_to_sized_spot(vehicle,
                                                     ParkingSpotSize.XL)
        elif isinstance(vehicle, Truck):
            parked_spot_id = \
                self._place_vehicle_to_sized_spot(vehicle, ParkingSpotSize.L)
            if parked_spot_id != -1:
                return parked_spot_id
            return self._place_vehicle_to_sized_spot(vehicle,
                                                     ParkingSpotSize.XL)
        else:
            return self._place_vehicle_to_sized_spot(vehicle,
                                                     ParkingSpotSize.XL)

    def _place_vehicle_to_sized_spot(self, vehicle: Vehicle,
                                     size: ParkingSpotSize) -> int:
        """
        Private helper method to place the given vehicle to a parking spot of
        the given size.
        :param vehicle: Vehicle
        :param size: ParkingSpotSize
        :return: int
        """
        size_available = self._available[size]
        if not size_available:
            return -1
        spot = size_available.popleft()
        spot.vehicle = vehicle
        self._parked[vehicle.license_plate] = spot
        return spot.id
        # Time: O(1)

    def retrieve_vehicle(self, license_plate: str) -> Vehicle:
        """
        Retrieves the vehicle with the given license plate.
        :param license_plate: str
        :return: Vehicle
        """
        # Check whether the vehicle is already parked
        if license_plate not in self._parked:
            raise Exception('The vehicle is not parked.')

        spot = self._parked.pop(license_plate)
        vehicle = spot.vehicle
        # Free the parking spot
        spot.vehicle = None
        size_available = self._available[spot.size]
        size_available.append(spot)
        self._available[spot.size] = size_available
        return vehicle
        # Time: O(1)

File: Interview-System Design Practice/Parking Lot/Java/parking_lot/ParkingLot.java
package parking_lot;

import vehicle.Bus;
import vehicle.Car;
import vehicle.MotorCycle;
import vehicle.Truck;
import vehicle.Vehicle;

import java.util.Arrays;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;

/**
 * ParkingLot class.
 *
 * @author Ziang Lu
 */
public class ParkingLot {

    /**
     * All parking spot sizes in ascending order.
     */
    private static final ParkingSpotSize[] ALL_SIZES_IN_ORDER =
            {ParkingSpotSize.S, ParkingSpotSize.M, ParkingSpotSize.M, ParkingSpotSize.XL};

    /**
     * Mapping between parking spot sizes and the corresponding available
     * parking spots.
     */
    private final Map<ParkingSpotSize, Queue<ParkingSpot>> available;
    /**
     * Mapping between the license plates of the parked vehicles and the
     * corresponding parking spot.
     */
    private final Map<String, ParkingSpot> parked;

    /**
     * Default constructor.
     */
    public ParkingLot() {
        available = new HashMap<>();
        for (ParkingSpotSize size : ParkingSpotSize.values()) {
            available.put(size, new ArrayDeque<>());
        }
        parked = new HashMap<>();
    }

    /**
     * Adds a parking spot of the given size to this parking lot.
     * @param size given parking spot size
     */
    public void addParkingSpot(ParkingSpotSize size) {
        available.get(size).offer(new ParkingSpot(size));
    }

    /**
     * Places the given vehicle.
     * @param v vehicle to place
     * @return parking spot ID on success, -1 on failure
     */
    public int placeVehicle(Vehicle v) {
        // Check whether the vehicle is already parked
        if (parked.containsKey(v.licensePlate())) {
            throw new RuntimeException("The vehicle is already parked.");
        }

        ParkingSpotSize vehicleSize = ParkingSpotSize.S;
        if (v instanceof Car) {
            vehicleSize = ParkingSpotSize.M;
        } else if (v instanceof Truck) {
            vehicleSize = ParkingSpotSize.L;
        } else if (v instanceof Bus) {
            vehicleSize = ParkingSpotSize.XL;
        }
        return placeVehicleFromSizedSpot(v, vehicleSize);
    }

    /**
     * Private helper method to place the given vehicle to a parking spot,
     * starting from searching the given size.
     * @param v vehicle to place
     * @param fromSize given parking spot size from which to start searching
     * @return parking spot ID on success, -1 on failure
     */
    private int placeVehicleFromSizedSpot(Vehicle v, ParkingSpotSize fromSize) {
        int fromIdx = Arrays.binarySearch(ALL_SIZES_IN_ORDER, fromSize);
        int parkedSpotId = -1;
        for (int idx = fromIdx; idx < ALL_SIZES_IN_ORDER.length; ++idx) {
            ParkingSpotSize spotSize = ALL_SIZES_IN_ORDER[idx];
            // Try to place the given vehicle to a parking spot of this size
            int parkedSpotId = placeVehicleToSizedSpot(v, spotSize);
            if (parkedSpotId != -1) {
                break;
            }
        }
        return parkedSpotId;
        // Time: O(1)
    }

    /**
     * Helper method to place the given vehicle to a parking spot of the given
     * size.
     * @param v vehicle to place
     * @param size given parking spot size
     * @return parking spot ID on success, -1 on failure
     */
    private int placeVehicleToSizedSpot(Vehicle v, ParkingSpotSize size) {
        Queue<ParkingSpot> sizeAvailable = available.get(size);
        if (sizeAvailable.isEmpty()) {
            return -1;
        }

        ParkingSpot spot = sizeAvailable.poll();
        // Occupy the parking spot
        spot.setVehicle(v);
        parked.put(v.licensePlate(), spot);
        return spot.id();
    }

    /**
     * Retrieves the vehicle with the given license plate.
     * @param licensePlate license plate to retrieve
     * @return retrieved vehicle
     */
    public Vehicle retrieveVehicle(String licensePlate) {
        // Check whether the vehicle is already parked
        if (!parked.containsKey(licensePlate)) {
            throw new RuntimeException("The vehicle is not parked.");
        }

        ParkingSpot spot = parked.remove(licensePlate);
        Vehicle v = spot.getVehicle();
        // Free the parking spot
        spot.setVehicle(null);
        available.get(spot.size()).offer(spot);
        return v;
    }

}
